dffml/source/dataframe.py

```diff
@@ -0,0 +1,71 @@
+"""
+Expose Pandas DataFrame as DFFML Source
+"""
+from collections import OrderedDict
+from typing import Dict, List, AsyncIterator
+
+
+from ..record import Record
+from ..base import config, field
+from ..util.entrypoint import entrypoint
+from .source import BaseSourceContext, BaseSource
+
+
+class DataFrameSourceContext(BaseSourceContext):
+    async def update(self, record: Record):
+
+        df = self.parent.config.dataframe
+
+        feature_columns = []
+        for col in df.columns:
+            feature_columns.append(col)
+
+        # Store feature data
+        feature_data = OrderedDict.fromkeys(feature_columns)
+        feature_data.update(record.features(feature_columns))
+        for col in feature_columns:
+            df.loc[record.key, col] = feature_data[col]
+
+        # Store prediction
+        try:
+            for col in self.parent.config.predictions:
+
+                prediction = record.prediction(col)
+                prediction_columns = self.parent.config.predictions
+                prediction_data = OrderedDict.fromkeys(prediction_columns)
+                prediction_data.update(prediction.dict())
+
+                df.loc[record.key, col] = prediction_data[col]
+
+        except KeyError:
```

Comment:
What is this about?

Reply:
This ensures that no changes are made to the records in case of a KeyError. This is more of a edge case scenario, so this can be removed if necessary.